File: Go/app/donp.go
package main

import (
	"encoding/json"
	"errors"
	"log/slog"
	"os"

	"donp/pkg/component"
	"donp/pkg/observe"
)

/*
	Welcome to the Desciptive Object Notated Protocol (DONP) application
    	where protocol messages described in JSON files are constructed,
		then used to simulate transactions via a simulated transport
		mechanism. The message log and application statistics are gathered
		and displayed at the end of the run!

		Copyright (c) 2025 Kathy Snell, All rights reserved.

*/

const (
	MODBUS_RTU   = "MODBUS_RTU"
	MODBUS_ASCII = "MODBUS_ASCII"
	ACTIVE       = MODBUS_RTU
)

// DescObjNotatedProtocolApp represents the Custom MTO Protocol application.
type DescObjNotatedProtocolApp struct {
	stats    observe.Statistics
	protocol *component.Protocol
}

// NewDescObjNotatedProtocolApp creates a new instance of DescObjNotatedProtocolApp, initializing statistics and logging.
func NewDescObjNotatedProtocolApp() *DescObjNotatedProtocolApp {
	app := &DescObjNotatedProtocolApp{}
	app.stats = *observe.NewStatistics()
	app.stats.StartTime()
	slog.SetLogLoggerLevel(slog.LevelInfo)
	slog.Info("Welcome to the DONP (Descriptive Object Notated Protocol) Application!")
	return app
}

// initProtocol uses data map based on JSON content, passing it on to the protocol for initialization.
func (app *DescObjNotatedProtocolApp) initProtocol() error {
	read, err := app.readFromJSON()
	if err != nil {
		return errors.New("DONP App: Unable to load protocol file")
	}
	var dataMap map[string]interface{}
	err = json.Unmarshal(read, &dataMap)
	if err != nil {
		return errors.New("DONP App: JSON error: " + err.Error())
	}
	if dataMap == nil {
		return errors.New("DONP App: Unmarshaled data map is nil")
	}
	// Create protocol from data map
	app.protocol = component.NewProtocol(dataMap)
	return nil
}

// readFromJSON reads the JSON configuration file based on the active protocol type.
func (app *DescObjNotatedProtocolApp) readFromJSON() ([]byte, error) {
	filePath := "../"
	switch ACTIVE {
	case MODBUS_RTU:
		filePath += "modbusRtu.json"
	case MODBUS_ASCII:
		filePath += "modbusAscii.json"
	default:
		return nil, errors.New("DONP App: Unknown protocol type: " + ACTIVE)
	}
	return os.ReadFile(filePath)
}

// run calls protocol Run method and finalizes statistics.
func (app *DescObjNotatedProtocolApp) Run() error {
	if app.protocol == nil {
		return errors.New("DONP App: Protocol not initialized, unable to run")
	}
	app.protocol.Run()
	app.stats.StopTime()
	app.stats.Log()
	return nil
}

func (app *DescObjNotatedProtocolApp) Log() {
	if app.protocol != nil {
		app.protocol.Log()
	}
}

func main() {
	app := NewDescObjNotatedProtocolApp()
	if err := app.initProtocol(); err != nil {
		slog.Error("DONP App: Failed to initialize protocol: " + err.Error())
		return
	}
	app.Log()
	app.Run()
}

File: Python/donp.py
import logging
from pkg.component.protocol import Protocol
import json
from pkg.observe.statistics import Statistics
from enum import Enum

class Protocols(Enum):    
    MODBUS_RTU = "MODBUS_RTU"
    MODBUS_ASCII = "MODBUS_ASCII"
    ACTIVE = MODBUS_RTU

"""
    Welcome to the Desciptive Object Notated Protocol (DONP) application
    	where protocol messages described in JSON files are constructed,
		then used to simulate transactions via a simulated transport
		mechanism. The message log and application statistics are gathered
		and displayed at the end of the run!

        Copyright (c) 2025 Kathy Snell, All rights reserved.

"""

class DescObjNotatedProtocolApp:

    def __init__(self):
        # Setup logger
        logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s', datefmt='%Y-%m-%d %H:%M:%S')    
        self.logger = logging.getLogger(__name__) 
        # Setup statistics
        self.stats = Statistics(self.logger)
        self.stats.start_time()
        self.logger.info("Welcome to the DONP (Descriptive Object Notated Protocol) Application!")
        self.protocol = None

    def log(self):
        """
        Logs (prints) the protocol object details.

        Args:
            None
        Returns:
            None
        Raises:
            None
        """
        if self.protocol is not None:
            self.protocol.log()

    def init_protocol(self):
        """
        Initializes the protocol by loading the protocol file.

        Args:
            None
        Returns:
            None
        Raises:
            None
        """
        file = self.open_file()
        try:
            if file is None:
                raise ValueError("DONP App: unable to load protocol file")
            self.protocol = Protocol(file)
        except Exception as e:
            self.logger.error("DONP App: an exception has occurred: %s", e)    

    def open_file(self):
        """
        Opens and loads the protocol JSON file.

        Args:
            None
        Returns:
            None
        Raises:
            None
        """
        file_path = "../"
        match Protocols.ACTIVE:
            case Protocols.MODBUS_ASCII:
                file_path = file_path + "modbusAscii.json"
            case Protocols.MODBUS_RTU:
                file_path = file_path + "modbusRtu.json"
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                return json.load(f)
        except Exception as e: 
            self.logger.error("DONP App: an exception has occurred: %s", e)
        return None
    
    def run(self):
        """
        Runs the protocol and reports statistics.

        Args:
            None
        Returns:
            None
        Raises:
            None
        """
        if self.protocol is not None:
            self.protocol.run()
            self.stats.stop_time()
            self.stats.log()
        else:
            self.logger.error("DONP App: Protocol is not initialized, unable to run.")

if __name__ == '__main__':
    """
        Initializes and runs the Custom MTO Protocol application.

        Args:
            None
        Returns:
            None
        Raises:
            None
        """
    app = DescObjNotatedProtocolApp()
    app.init_protocol()
    app.log()
    app.run()
